### ISSUE TITLE
Concurrency issue in generating tensor name in newTensor

### ISSUE THREAD

**Issue by pjongy** [sugarme/gotch#110]
In parallelized environment I think generating name without wrapped by lock/unlock block can cause duplicated tensor name to be generated.

For current lock/unlock block for preventing concurrently modifying about `ExistsingTensor` is pretty good, but about name of tensor is not.
https://github.com/sugarme/gotch/blob/cd0565716fd369786b0fe016e38c46ba4046bf37/ts/tensor.go#L59-L72

And if duplicated name exists, it can be dangling pointer because of not free-ed by `freeTensor`
https://github.com/sugarme/gotch/blob/cd0565716fd369786b0fe016e38c46ba4046bf37/ts/tensor.go#L141-L145

**Comment by pjongy** [comment]
This is my suggestion (lock and defer like freeTensor's`):
as-is:
```
func newTensor(ctensor lib.Ctensor, nameOpt ...string) *Tensor {
	if len(nameOpt) == 0 {
		nameOpt = []string{}
	}
	name := newName(nameOpt...)

	x := new(Tensor)
	x.ctensor = ctensor
	x.d = new(bigStruct)

	atomic.AddInt64(&TensorCount, 1)
	nbytes := x.nbytes()
	atomic.AddInt64(&AllocatedMem, nbytes)
	lock.Lock()
	if _, ok := ExistingTensors[name]; ok {
		name = fmt.Sprintf("%s_%09d", name, TensorCount)
	}
	ExistingTensors[name] = struct{}{}
	lock.Unlock()
```

to-be:

```
func newTensor(ctensor lib.Ctensor, nameOpt ...string) *Tensor {
	if len(nameOpt) == 0 {
		nameOpt = []string{}
	}
	lock.Lock()
	defer lock.Unlock()
	name := newName(nameOpt...)

	x := new(Tensor)
	x.ctensor = ctensor
	x.d = new(bigStruct)

	atomic.AddInt64(&TensorCount, 1)
	nbytes := x.nbytes()
	atomic.AddInt64(&AllocatedMem, nbytes)
	if _, ok := ExistingTensors[name]; ok {
		name = fmt.Sprintf("%s_%09d", name, TensorCount)
	}
	ExistingTensors[name] = struct{}{}
```

**Comment by sugarme** [comment]
@pjongy ,

Thanks for comments. Placing `newName()` in lock block looks good and easy to read although I have tested concurrently with `-race` flag without any issues. Feel free to PR (would love to have some unit tests with `-race` flag so that we don't break something).

I had tried to name tensors with timestamp however, when it comes to debugging and tracing tensors/operations when having issues, simple naming seems to be much easier.

Cheers!
